src/model/dfcir/lib/dfcir/conversions/FIRRTLStubsGeneratorPass.cpp

```diff
@@ -0,0 +1,164 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the Utopia HLS Project, under the Apache License v2.0
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2021-2024 ISP RAS (http://www.ispras.ru)
+//
+//===----------------------------------------------------------------------===//
+
+#include "dfcir/conversions/DFCIRPasses.h"
+#include "dfcir/conversions/DFCIRPassesUtils.h"
+
+#include "circt/Dialect/FIRRTL/FIRRTLDialect.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "ctemplate/template.h"
+#include "mlir/IR/BuiltinOps.h"
+
+#include <algorithm>
+#include <ctime>
+#include <string>
+
+namespace mlir::dfcir {
+
+#define GEN_PASS_DECL_FIRRTLSTUBGENERATORPASS
+#define GEN_PASS_DEF_FIRRTLSTUBGENERATORPASS
+
+#include "dfcir/conversions/DFCIRPasses.h.inc"
+
+class FIRRTLStubGeneratorPass
+    : public impl::FIRRTLStubGeneratorPassBase<FIRRTLStubGeneratorPass> {
+  using TemplateDictionary = ctemplate::TemplateDictionary;
+  using FExtModuleOp = circt::firrtl::FExtModuleOp;
+  using CircuitOp = circt::firrtl::CircuitOp;
+  using FIRRTLBaseType = circt::firrtl::FIRRTLBaseType;
+  using IntType = circt::firrtl::IntType;
+
+private:
+  TemplateDictionary *processFIFOModule(TemplateDictionary *dict,
+                                        FExtModuleOp module) {
+    TemplateDictionary *result = dict->AddSectionDictionary("FIFO_MODULES");
+    auto ports = module.getPorts();
+    auto res1Type = llvm::cast<FIRRTLBaseType>(module.getPortType(0));
+    int32_t width = res1Type.getBitWidthOrSentinel();
+    result->SetFormattedValue("WIDTH", "%d", width - 1);
+    result->SetValue("RES1", ports[0].getName().data());
+    result->SetValue("ARG1", ports[1].getName().data());
+    result->SetValue("CLK", ports[2].getName().data());
+    return result;
+  }
+
+  TemplateDictionary *processBinModule(TemplateDictionary *dict,
+                                       FExtModuleOp module,
+                                       unsigned latency) {
+    TemplateDictionary *result = dict->AddSectionDictionary("BINARY_MODULES");
+    auto moduleName = module.getModuleName();
+    if (moduleName.contains(ADD_MODULE)) {
+      result->SetValue("OP", "+");
+    } else if (moduleName.contains(SUB_MODULE)) {
+      result->SetValue("OP", "-");
+    } else {
+      result->SetValue("OP", "*");
+    }
```

Comment:
Add separate branch for MUL operation, report an error ("Unsupported module: " ?) at the default branch